src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettingsTemplatePackageProvider.cs

```diff
@@ -305,13 +306,53 @@ private async Task<InstallResult> InstallAsync(List<TemplatePackageData> package
                 throw new InvalidOperationException($"{nameof(installResult.TemplatePackage)} cannot be null when {nameof(installResult.Success)} is 'true'");
             }
 
+            // We check this after installation to make sure that we captured all templates inside a package
+            var (code, errorMessage) = await EnsureUniqueIdentifierAsync(installResult, installRequest.Force, cancellationToken).ConfigureAwait(false);
+
+            if (code == InstallerErrorCode.DuplicatedIdentity)
+            {
+                await installer.UninstallAsync(installResult.TemplatePackage, this, cancellationToken).ConfigureAwait(false);
+                return InstallResult.CreateFailure(installRequest, code, errorMessage);
+            }
+
             lock (packages)
             {
                 packages.Add(((ISerializableInstaller)installer).Serialize(installResult.TemplatePackage));
             }
             return installResult;
         }
 
+        private async Task<(InstallerErrorCode, string)> EnsureUniqueIdentifierAsync(InstallResult installResult, bool force, CancellationToken cancellationToken)
+        {
+            if (installResult.TemplatePackage is null)
+            {
+                throw new InvalidOperationException($"No package found in the result of this install operation");
+            }
+
+            var scanResult = new Scanner(_environmentSettings).Scan(installResult.TemplatePackage.MountPointUri, scanForComponents: false);
+
+            var packageManager = new TemplatePackageManager(_environmentSettings);
+            var installedTemplates = await packageManager.GetTemplatesAsync(cancellationToken).ConfigureAwait(false);
+
+            foreach (ITemplate template in scanResult.Templates)
+            {
+                if (installedTemplates.FirstOrDefault(s => s.Identity == template.Identity) is ITemplateInfo duplicatedTemplate)
+                {
+                    var managedPackage = await packageManager.GetTemplatePackageAsync(template, cancellationToken).ConfigureAwait(false);
+
+                    if (managedPackage.MountPointUri.Equals(duplicatedTemplate.MountPointUri, StringComparison.OrdinalIgnoreCase) || !force)
+                    {
+                        return (InstallerErrorCode.DuplicatedIdentity, string.Format(LocalizableStrings.GlobalSettingsTemplatePackageProvider_InstallResult_Error_DuplicatedIdentity, duplicatedTemplate.Name));
+                    }
+
+                    _logger.LogWarning(message: $"The template {installResult.TemplatePackage} has the identity conflict with an already existing template {duplicatedTemplate.Name}. " +
+                        $"The template {duplicatedTemplate.Name} will be overwritten.");
```

Comment:
can we log in English directly like this, or does it need to be a localizable string?

Reply:
warnings should be localized